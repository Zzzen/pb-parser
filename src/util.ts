export function assertIsDefined<T>(
  value: T,
  message: string
): asserts value is NonNullable<T> {
  // eslint-disable-next-line no-null/no-null
  if (value === undefined || value === null) {
    throw new Error(message);
  }
}
