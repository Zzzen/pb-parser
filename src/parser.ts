import { Scanner, Token, TokenType, Keyword } from "./scanner";
import {
  Comment,
  SyntaxStatement,
  ProtoFile,
  TopLevelDirective,
  ImportStatement,
  Package,
  FullIdentifier,
  Option,
  OptionName,
  Constant,
  TopLevelDef,
  Message,
  Reserved,
  MessageBody,
  MapField,
  ValueType,
} from "./ast";
import { last } from "./util";

export class Parser {
  pendingComments: Comment[] = [];
  current = 0;
  tokens: Token[] = [];

  constructor(private readonly source: string) {}

  handleComments() {
    while (this.check(TokenType.COMMENT)) {
      const token = this.peek();
      this.pendingComments.push({
        type: "Comment",
        value: token.literal as string,
        start: token.start,
        end: token.end,
      });
    }
  }

  parse() {
    this.tokens = new Scanner(this.source).scanTokens();
    return this.parseRoot();
  }

  parseSyntax(): SyntaxStatement | undefined {
    this.handleComments();
    if (this.checkKeyword(Keyword.SYNTAX)) {
      const syntax = this.advance();
      this.handleComments();
      this.consume(TokenType.EQUAL);
      const proto = this.consume(TokenType.STRING);
      const semicolon = this.consume(TokenType.SEMICOLON);
      return {
        type: "SyntaxStatement",
        value: proto.literal as "proto2" | "proto3",
        start: syntax.start,
        end: semicolon.end,
      };
    }
  }

  parseTopLevelDirectives(): TopLevelDirective[] {
    const topLevelDirectives: TopLevelDirective[] = [];
    while (!this.isAtEnd()) {
      if (this.checkKeyword(Keyword.IMPORT)) {
        topLevelDirectives.push(this.parseImport());
        continue;
      } else if (this.checkKeyword(Keyword.PACKAGE)) {
        topLevelDirectives.push(this.parsePackage());
        continue;
      } else if (this.checkKeyword(Keyword.OPTION)) {
        topLevelDirectives.push(this.parseOption());
        continue;
      } else if (this.match(TokenType.SEMICOLON)) {
        continue;
      } else if (
        this.peek().type === TokenType.IDENTIFIER &&
        [
          Keyword.MESSAGE,
          Keyword.ENUM,
          Keyword.EXTEND,
          Keyword.SERVICE,
        ].includes(this.peek().lexeme as any)
      ) {
        topLevelDirectives.push(this.parseTopLevelDef());
        continue;
      }

      this.error(this.peek(), `unexpected token`);
    }

    return topLevelDirectives;
  }

  /**
   * topLevelDef = message | enum | extend | service
   */
  parseTopLevelDef(): TopLevelDef {
    if (this.checkKeyword(Keyword.MESSAGE)) {
      return this.parseMessage();
    }

    this.error(this.peek(), "");
  }

  /**
   * message = "message" messageName messageBody
   * messageBody = "{" { field | enum | message | extend | extensions | group |
   *   option | oneof | mapField | reserved | emptyStatement } "}"
   */
  parseMessage(): Message {
    const messageToken = this.consumeKeyword(Keyword.MESSAGE);
    const name = this.consume(TokenType.IDENTIFIER).lexeme as string;
    this.consume(TokenType.LEFT_BRACE);
    const body = this.parseMessageBody();
    const end = this.consume(TokenType.RIGHT_BRACE);

    return {
      type: "Message",
      start: messageToken.start,
      end: end.end,
      body,
      name,
    };
  }

  parseMessageBody(): MessageBody[] {
    const body: MessageBody[] = [];
    while (!this.check(TokenType.RIGHT_BRACE)) {
      if (this.checkKeyword(Keyword.RESERVED)) {
        body.push(this.parserReserved());
        continue;
      } else if (this.checkKeyword(Keyword.MAP)) {
        body.push(this.parseMapField());
        continue;
      } else if (this.checkKeyword(Keyword.OPTION)) {
        body.push(this.parseOption());
        continue;
      }

      // emptyStatement
      if (this.match(TokenType.SEMICOLON)) {
        continue;
      }

      this.error(this.peek());
    }
    return body;
  }

  /**
   * mapField = "map" "<" keyType "," type ">" mapName "=" fieldNumber [ "[" fieldOptions "]" ] ";"
   * keyType = "int32" | "int64" | "uint32" | "uint64" | "sint32" | "sint64" |
   *          "fixed32" | "fixed64" | "sfixed32" | "sfixed64" | "bool" | "string"
   */

  parseMapField(): MapField {
    const start = this.consumeKeyword(Keyword.MAP);
    this.consume(TokenType.LESS);
    // keyType
    const keyToken = this.consume(TokenType.PRIMITIVE_TYPE);
    this.consume(TokenType.COMMA);

    const value = this.parseValueType();

    this.consume(TokenType.GREATER);

    const id = this.consume(TokenType.IDENTIFIER);

    this.consume(TokenType.EQUAL);

    const numberToken = this.consume(TokenType.NUMBER);

    let options: Option[] = [];
    if (this.match(TokenType.LEFT_BRACKET)) {
      options = this.parseFieldOptions();
      this.consume(TokenType.RIGHT_BRACKET);
    }

    const end = this.consume(TokenType.SEMICOLON);

    return {
      type: "MapField",
      start: start.start,
      end: end.end,
      valueType: value,
      keyType: keyToken.lexeme,
      mapName: id.lexeme,
      fieldNumber: Number(numberToken.lexeme),
      options,
    };
  }

  /**
   * type = "double" | "float" | "int32" | "int64" | "uint32" | "uint64"
   *   | "sint32" | "sint64" | "fixed32" | "fixed64" | "sfixed32" | "sfixed64"
   *   | "bool" | "string" | "bytes" | messageType | enumType
   *
   *
   * messageType = [ "." ] { ident "." } messageName
   * enumType = [ "." ] { ident "." } enumName
   */
  parseValueType(): ValueType {
    if (this.match(TokenType.PRIMITIVE_TYPE)) {
      const token = this.previous();
      return {
        type: "ValueType",
        start: token.start,
        end: token.end,
        value: token.lexeme,
      };
    } else {
      const leadingDot = this.check(TokenType.DOT) ? this.advance() : undefined;
      const idents = [] as Token[];
      while (this.check(TokenType.IDENTIFIER)) {
        idents.push(this.advance());
        if (!this.match(TokenType.DOT)) {
          break;
        }
      }

      return {
        type: "ValueType",
        start: leadingDot ? leadingDot.start : idents[0].start,
        end: last(idents).end,
        value:
          (leadingDot ? "." : "") + idents.map((id) => id.lexeme).join("."),
      };
    }
  }

  /**
   * reserved = "reserved" ( ranges | fieldNames ) ";"
   * ranges = range { "," range }
   * range =  intLit [ "to" ( intLit | "max" ) ]
   * fieldNames = fieldName { "," fieldName }
   */
  parserReserved(): Reserved {
    const reserved = this.consumeKeyword(Keyword.RESERVED);
    if (this.check(TokenType.STRING)) {
      // fieldNames
      const ids = [this.advance()];
      while (this.match(TokenType.COMMA)) {
        ids.push(this.consume(TokenType.STRING));
      }

      const end = this.consume(TokenType.SEMICOLON);
      return {
        type: "Reserved",
        ranges: undefined,
        fieldName: ids.map((x) => x.literal as string),
        start: reserved.start,
        end: end.end,
      };
    } else {
      // ranges
      const ranges = [this.parseRange()];
      while (this.match(TokenType.COMMA)) {
        ranges.push(this.parseRange());
      }

      const end = this.consume(TokenType.SEMICOLON);
      return {
        type: "Reserved",
        ranges: ranges,
        fieldName: undefined,
        start: reserved.start,
        end: end.end,
      };
    }
  }

  parseRange(): [number, number | "max"] | [number] {
    const lit = this.consume(TokenType.NUMBER);
    const value = lit.literal as number;
    if (this.matchKeyword(Keyword.TO)) {
      if (this.matchKeyword(Keyword.MAX)) {
        return [value, "max"];
      } else {
        const toValue = this.consume(TokenType.NUMBER).literal as number;
        return [value, toValue];
      }
    } else {
      return [value];
    }
  }

  /**
   * option = "option" optionName  "=" constant ";"
   */
  parseOption(): Option {
    const start = this.consumeKeyword(Keyword.OPTION);

    const optionName = this.parseOptionName();

    this.consume(TokenType.EQUAL);

    const constant = this.parseConstant();

    const end = this.consume(TokenType.SEMICOLON);

    return {
      type: "Option",
      name: optionName,
      value: constant,
      start: start.start,
      end: end.end,
    };
  }

  /**
   * fieldOptions = fieldOption { ","  fieldOption }
   */
  parseFieldOptions() {
    const opts = [this.parseFieldOption()];
    while (this.match(TokenType.COMMA)) {
      opts.push(this.parseFieldOption());
    }

    return opts;
  }

  /**
   * fieldOption = optionName "=" constant
   */
  parseFieldOption(): Option {
    const name = this.parseOptionName();
    this.consume(TokenType.EQUAL);
    const constant = this.parseConstant();
    return {
      type: "Option",
      start: name.start,
      end: constant.end,
      value: constant,
      name: name,
    };
  }

  /**
   * optionName = ( ident | "(" fullIdent ")" ) { "." ident }
   */
  parseOptionName(): OptionName {
    let name = "";
    const start = this.peek();
    if (this.check(TokenType.IDENTIFIER)) {
      name = this.consume(TokenType.IDENTIFIER).lexeme;
    } else {
      this.consume(TokenType.LEFT_PAREN);
      name += "(";

      const fullId = this.parseFullIdentifier();
      name += fullId.value;

      this.consume(TokenType.RIGHT_PAREN);
      name += ")";
    }

    while (this.match(TokenType.DOT)) {
      name += ".";
      name += this.consume(TokenType.IDENTIFIER).lexeme;
    }

    return {
      type: "OptionName",
      value: name,
      start: start.start,
      end: this.previous().end,
    };
  }

  /**
   * constant = fullIdent | ( [ "-" | "+" ] intLit ) | ( [ "-" | "+" ] floatLit ) |
   *                 strLit | boolLit
   */
  parseConstant(): Constant {
    if (this.check(TokenType.IDENTIFIER)) {
      return this.parseFullIdentifier();
    } else {
      let symbol: Token;
      if (this.check(TokenType.MINUS) || this.check(TokenType.PLUS)) {
        symbol = this.advance();
        const isMinue = symbol.type === TokenType.MINUS;
        return {
          type: "Literal",
          value:
            (isMinue ? -1 : +1) *
            (this.consume(TokenType.NUMBER).literal as number),
          start: symbol.start,
          end: symbol.end,
        };
      } else {
        const token = this.consume(TokenType.STRING);
        return {
          type: "Literal",
          value: token.literal as string,
          start: token.start,
          end: token.end,
        };
      }
    }
  }

  parsePackage(): Package {
    const packageToken = this.consumeKeyword(Keyword.PACKAGE);

    const name = this.parseFullIdentifier();

    const end = this.consume(TokenType.SEMICOLON);

    return {
      type: "Package",
      name,
      start: packageToken.start,
      end: end.end,
    };
  }

  parseFullIdentifier(): FullIdentifier {
    const ids = [this.consume(TokenType.IDENTIFIER)];
    while (this.match(TokenType.DOT)) {
      ids.push(this.consume(TokenType.IDENTIFIER));
    }

    const fullId = ids.map((id) => id.lexeme).join(".");

    return {
      type: "FullIdentifier",
      value: fullId,
      start: ids[0].start,
      end: ids[ids.length - 1].end,
    };
  }

  parseImport(): ImportStatement {
    const importToken = this.consumeKeyword(Keyword.IMPORT);
    let modifier: Token | undefined;
    if (this.checkKeyword(Keyword.WEAK) || this.checkKeyword(Keyword.PUBLIC)) {
      modifier = this.advance();
    }

    const strLit = this.consume(TokenType.STRING);

    const end = this.consume(TokenType.SEMICOLON);

    return {
      type: "ImportStatement",
      file: strLit.literal as string,
      modifier: modifier?.lexeme as "weak" | "public",
      start: importToken.start,
      end: end.end,
    };
  }

  /**
   * proto = syntax { import | package | option | topLevelDef | emptyStatement }
   */
  parseRoot(): ProtoFile {
    const syntax = this.parseSyntax();
    const body = this.parseTopLevelDirectives();
    return {
      type: "ProtoFile",
      syntax,
      body,
      start: this.tokens[0].start,
      end: this.tokens[this.tokens.length - 1].end,
    };
  }

  private previous() {
    return this.tokens[this.current - 1];
  }

  check(type: TokenType) {
    if (this.isAtEnd()) return false;
    return this.peek().type == type;
  }

  checkKeyword(type: Keyword) {
    if (this.isAtEnd()) return false;
    return (
      this.peek().type == TokenType.IDENTIFIER && this.peek().lexeme === type
    );
  }

  checkNext(type: TokenType) {
    return this.tokens[this.current + 1]?.type === type;
  }

  match(...types: TokenType[]) {
    for (const type of types) {
      if (this.check(type)) {
        this.advance();
        return true;
      }
    }

    return false;
  }

  matchKeyword(...types: Keyword[]) {
    for (const type of types) {
      if (this.checkKeyword(type)) {
        this.advance();
        return true;
      }
    }

    return false;
  }

  private consume(type: TokenType) {
    if (this.check(type)) return this.advance();

    const current = this.tokens[this.current];
    const errorMessage = `expected ${type}, but got ${current.type} at line ${current.line}`;

    throw this.error(this.peek(), errorMessage);
  }

  private consumeKeyword(type: Keyword) {
    if (this.checkKeyword(type)) return this.advance();

    const current = this.tokens[this.current];
    const errorMessage = `expected ${type}, but got ${current.type} at line ${current.line}`;

    throw this.error(this.peek(), errorMessage);
  }

  private error(token: Token, message?: string): never {
    throw new Error(
      `parsing Error: unexpect token ${token.type}(${token.lexeme}), ${message}`
    );
  }

  private advance() {
    if (!this.isAtEnd()) this.current++;
    return this.previous();
  }

  private isAtEnd() {
    return this.peek().type === TokenType.EOF;
  }

  private peek() {
    return this.tokens[this.current];
  }
}
