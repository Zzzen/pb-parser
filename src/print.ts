import { ASTNode, Extensions } from "./ast";
import { UnreachableCaseError } from "./util";

export function print(node: ASTNode | undefined): string {
  return str(node, 0);

  function str(
    node: ASTNode | undefined,
    indent: number,
    opt?: { expression: boolean }
  ) {
    if (node == undefined) {
      return "";
    }

    const isExpression = opt?.expression;
    const EOL = "\n";

    let output = "";
    switch (node.type) {
      case "ProtoFile":
        writeText(str(node.syntax, indent));
        writeStrNodes(node.body, indent);
        break;
      case "Comment":
        writeWithIndent(`/* ${node.value} */` + EOL);
        break;
      case "SyntaxStatement":
        writeWithIndent(`syxtax = "${node.value}";` + EOL);
        break;
      case "ImportStatement":
        writeWithIndent(`import ${node.modifier || ""} "${node.file}";` + EOL);
        break;
      case "Option":
        writeWithIndent(
          `option ${str(node.name, 0)} = ${str(node.value, 0)}${
            isExpression ? "" : ";\n"
          }`
        );
        break;
      case "Package":
        writeWithIndent(`package ${str(node.name, 0)};` + EOL);
        break;
      case "FullIdentifier":
        writeText(`${node.value}`);
        break;
      case "OptionName":
        writeText(`${node.value}`);
        break;
      case "Literal":
        writeText(JSON.stringify(node.value));
        break;
      case "Message":
        writeWithIndent(`message ${node.name} {` + EOL);
        writeStrNodes(node.body, indent + 1);
        writeWithIndent(`}`);
        break;
      case "Enum":
        writeWithIndent(`enum ${node.name} {` + EOL);
        writeStrNodes(node.options, indent + 1);
        writeStrNodes(node.fields, indent + 1);
        writeWithIndent(`}`);
        break;
      case "Extend":
        throw new Error(`unsupported type`);
      case "Service":
        writeWithIndent(`service ${node.name} {` + EOL);
        writeStrNodes(node.body, indent + 1);
        writeWithIndent(`}`);
        break;
      case "Field":
        const hasOption = node.fieldOptions.length > 0;
        const options = hasOption
          ? `[${node.fieldOptions.map(
              (field) => `${str(field, 0, { expression: true })}`
            )}]`
          : "";
        writeWithIndent(
          `${node.label || ""}  ${str(node.fieldType, 0)} ${node.name} = ${
            node.fieldNumber
          } ${options};` + EOL
        );
        break;
      case "Extensions":
        writeWithIndent(`extensions ${formatRange(node.ranges)} ;` + EOL);
        break;
      case "Group":
        throw new Error(`unsupported type`);
      case "OneOf":
        writeWithIndent(`oneof ${node.oneofName}` + EOL);
        writeStrNodes(node.options, indent + 1);
        writeStrNodes(node.fields, indent + 1);
        writeWithIndent(`}`);
        break;
      case "MapField":
        const hasMapOption = node.options.length > 0;
        const mapOptions = hasMapOption
          ? `[${node.options.map(
              (field) => `${str(field, 0, { expression: true })}`
            )}]`
          : "";
        writeWithIndent(
          `map<${node.keyType}, ${str(node.valueType, 0, {
            expression: true,
          })}> ${node.mapName} = ${node.fieldNumber} ${mapOptions};` + EOL
        );
        break;
      case "ValueType":
        writeText(node.value);
        break;
      case "Reserved":
        writeWithIndent(
          `reserved ${formatRange(node.ranges)} ${node.fieldName?.join(", ")};`
        );
        break;
      case "RPC":
        const hasRPCOptions = node.options.length > 0;
        writeWithIndent(
          `rpc ${node.name} (${node.streamInput ? "stream" : ""} ${str(
            node.input,
            0
          )}) returns (${node.streamOutput ? "stream" : ""} ${str(
            node.output,
            0
          )})${hasRPCOptions ? "{" : ";"}` + EOL
        );
        if (hasRPCOptions) {
          writeStrNodes(node.options, indent + 1);
          writeWithIndent(`}`);
        }
        break;
      case "Service":
        writeWithIndent(`service ${node.name} {` + EOL);
        writeStrNodes(node.body, indent + 1);
        writeWithIndent(`}`);
        break;
      default:
        throw new UnreachableCaseError(node);
    }

    return output;

    function writeStrNodes(nodes: ASTNode[], indent: number) {
      writeText(nodes.map((node) => str(node, indent)).join("\n"));
    }

    function writeWithIndent(t: string) {
      output += "\t".repeat(indent) + t;
    }
    function writeText(t: string) {
      output += t;
    }

    function formatRange(ranges: Extensions["ranges"]) {
      if (!ranges) {
        throw new Error("unexpected range");
      }

      if (ranges.length === 0) {
        return "";
      }

      return ranges
        .map((range) => {
          if (range.length === 1) {
            return range[0];
          } else {
            return range.join(" to ");
          }
        })
        .join(", ");
    }
  }
}
