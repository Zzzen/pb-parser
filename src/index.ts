import { Parser } from "./parser";

export function parse(src: string, _options: unknown) {
  return new Parser(src).parse();
}

export * from "./parser";
export * from "./scanner";
export * from "./ast";
