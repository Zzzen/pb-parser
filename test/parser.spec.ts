import { Parser } from "../src/parser";
import { print } from "../src/print";

function parse(src: string) {
  return new Parser(src).parse();
}

describe("Parser", () => {
  it("parse syntax", () => {
    expect(parse('syntax = "proto2";')).toMatchSnapshot();
  });

  it("parse import", () => {
    expect(
      parse(`
import "./test.proto";
import weak "./test.proto";
import public "./test.proto";
    `)
    ).toMatchSnapshot();
  });

  it("parse package", () => {
    expect(
      parse(`
package a;
package a.bc;
    `)
    ).toMatchSnapshot();
  });

  it("parse option", () => {
    expect(
      parse(`
option java_package = "com.example.foo";
option java_package = -1;
option java_package = "com.example.foo";
option (java._package) = "com.example.foo";
option java.pp.cc = "com.example.foo";
`)
    ).toMatchSnapshot();
  });

  it("parse empty statment", () => {
    expect(
      parse(`
;;;;;
    `)
    ).toMatchSnapshot();
  });

  it("parse reserved", () => {
    expect(
      parse(`
message M {
  reserved 1;
  reserved 1 to 2, 5 to max, 6;
  reserved "foo";
  reserved "foo", "bar";
}
      
      `)
    ).toMatchSnapshot();
  });

  it("parse map field", () => {
    expect(
      parse(`
message M {
  map <int32, string> f1 = 1;
  map <int32, M> f2 = 2;
  map <int32, M.foo> f3 = 3;
  map <int32, .M> f4 = 4;
  map <int32, .M.foo.bar> f5 = 5 [(custom_option) = "hello world"];
  map <int32, .M> map = 4;
}      
      
      `)
    ).toMatchSnapshot();
  });

  it("parse option field", () => {
    expect(
      parse(`
message M {
  option (my_option).a = true;
}      
      
      `)
    ).toMatchSnapshot();
  });

  it("parse oneof field", () => {
    expect(
      parse(`
message M {
  oneof foo {
      string name = 4;
      SubMessage sub_message = 9;
  }
}

      `)
    ).toMatchSnapshot();
  });

  it("parse nested message", () => {
    expect(
      parse(`
message M {
  message Inner {
    option (my_option).a = true;
  }
}

      `)
    ).toMatchSnapshot();
  });

  it("parse enum", () => {
    expect(
      parse(`
// this is enum
enum EnumAllowingAlias {
  option /** inline comments **/ allow_alias = true;  // this is an option
  UNKNOWN = 0;                // this is an enum field
  STARTED = 1;
  RUNNING = 2 [(custom_option) = "hello world"];
}

message M {
  enum EnumAllowingAlias {
    option allow_alias = true;
    UNKNOWN = 0;
    STARTED = 1;
    RUNNING = 2 [(custom_option) = "hello world"];
  }
}
      `)
    ).toMatchSnapshot();
  });

  it("parse extensions", () => {
    expect(
      parse(`
message M {
  extensions 20 to 30;
}
      `)
    ).toMatchSnapshot();
  });

  it("parse normal fields", () => {
    expect(
      parse(`
message M {
  optional foo.bar nested_message = 2;
  repeated int32 samples = 4 [packed=true];
  foo.bar nested_message = 2;
}
      `)
    ).toMatchSnapshot();
  });

  it("parse service", () => {
    expect(
      parse(`
service SearchService {
  rpc Search (SearchRequest) returns (SearchResponse);
}
      `)
    ).toMatchSnapshot();
  });

  it("parse example proto2", () => {
    expect(
      parse(`
syntax = "proto2";
import public "other.proto";
option java_package = "com.example.foo";
enum EnumAllowingAlias {
  option allow_alias = true;
  UNKNOWN = 0;
  STARTED = 1;
  RUNNING = 2 [(custom_option) = "hello world"];
}
message outer {
  option (my_option).a = true;
  message inner {
    required int64 ival = 1;
  }
  repeated inner inner_message = 2;
  optional EnumAllowingAlias enum_field = 3;
  map<int32, string> my_map = 4;
  extensions 20 to 30;
}
    `)
    ).toMatchSnapshot();
  });

  it("parse example proto3", () => {
    const ast = parse(`
syntax = "proto3";
import public "other.proto";
option java_package = "com.example.foo";
enum EnumAllowingAlias {
  option allow_alias = true;
  UNKNOWN = 0;
  STARTED = 1;
  RUNNING = 2 [(custom_option) = "hello world"];
}
message outer {
  option (my_option).a = true;
  message inner {
    int64 ival = 1;
  }
  repeated inner inner_message = 2;
  EnumAllowingAlias enum_field =3;
  map<int32, string> my_map = 4;
}
    `);
    expect(ast).toMatchSnapshot();
    expect(print(ast)).toMatchSnapshot();
  });
});
