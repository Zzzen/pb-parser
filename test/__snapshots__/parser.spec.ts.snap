// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser parse empty statment 1`] = `
Object {
  "body": Array [],
  "end": 11,
  "start": 1,
  "syntax": undefined,
  "type": "ProtoFile",
}
`;

exports[`Parser parse example proto 1`] = `
Object {
  "body": Array [
    Object {
      "end": 48,
      "file": "other.proto",
      "modifier": "public",
      "start": 20,
      "type": "ImportStatement",
    },
    Object {
      "end": 89,
      "name": Object {
        "end": 68,
        "start": 56,
        "type": "OptionName",
        "value": "java_package",
      },
      "start": 49,
      "type": "Option",
      "value": Object {
        "end": 88,
        "start": 71,
        "type": "Literal",
        "value": "com.example.foo",
      },
    },
  ],
  "end": 94,
  "start": 1,
  "syntax": Object {
    "end": 19,
    "start": 1,
    "type": "SyntaxStatement",
    "value": "proto2",
  },
  "type": "ProtoFile",
}
`;

exports[`Parser parse import 1`] = `
Object {
  "body": Array [
    Object {
      "end": 23,
      "file": "./test.proto",
      "modifier": undefined,
      "start": 1,
      "type": "ImportStatement",
    },
    Object {
      "end": 51,
      "file": "./test.proto",
      "modifier": "weak",
      "start": 24,
      "type": "ImportStatement",
    },
    Object {
      "end": 81,
      "file": "./test.proto",
      "modifier": "public",
      "start": 52,
      "type": "ImportStatement",
    },
  ],
  "end": 86,
  "start": 1,
  "syntax": undefined,
  "type": "ProtoFile",
}
`;

exports[`Parser parse map field 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "end": 42,
          "fieldNumber": 1,
          "keyType": "int32",
          "mapName": "f1",
          "options": Array [],
          "start": 15,
          "type": "MapField",
          "valueType": Object {
            "end": 33,
            "start": 27,
            "type": "ValueType",
            "value": "string",
          },
        },
        Object {
          "end": 67,
          "fieldNumber": 2,
          "keyType": "int32",
          "mapName": "f2",
          "options": Array [],
          "start": 45,
          "type": "MapField",
          "valueType": Object {
            "end": 58,
            "start": 57,
            "type": "ValueType",
            "value": "M",
          },
        },
        Object {
          "end": 96,
          "fieldNumber": 3,
          "keyType": "int32",
          "mapName": "f3",
          "options": Array [],
          "start": 70,
          "type": "MapField",
          "valueType": Object {
            "end": 87,
            "start": 82,
            "type": "ValueType",
            "value": "M.foo",
          },
        },
        Object {
          "end": 122,
          "fieldNumber": 4,
          "keyType": "int32",
          "mapName": "f4",
          "options": Array [],
          "start": 99,
          "type": "MapField",
          "valueType": Object {
            "end": 113,
            "start": 111,
            "type": "ValueType",
            "value": ".M",
          },
        },
        Object {
          "end": 190,
          "fieldNumber": 5,
          "keyType": "int32",
          "mapName": "f5",
          "options": Array [
            Object {
              "end": 188,
              "name": Object {
                "end": 172,
                "start": 157,
                "type": "OptionName",
                "value": "(custom_option)",
              },
              "start": 157,
              "type": "Option",
              "value": Object {
                "end": 188,
                "start": 175,
                "type": "Literal",
                "value": "hello world",
              },
            },
          ],
          "start": 125,
          "type": "MapField",
          "valueType": Object {
            "end": 147,
            "start": 137,
            "type": "ValueType",
            "value": ".M.foo.bar",
          },
        },
        Object {
          "end": 217,
          "fieldNumber": 4,
          "keyType": "int32",
          "mapName": "map",
          "options": Array [],
          "start": 193,
          "type": "MapField",
          "valueType": Object {
            "end": 207,
            "start": 205,
            "type": "ValueType",
            "value": ".M",
          },
        },
      ],
      "end": 219,
      "name": "M",
      "start": 1,
      "type": "Message",
    },
  ],
  "end": 239,
  "start": 1,
  "syntax": undefined,
  "type": "ProtoFile",
}
`;

exports[`Parser parse oneof field 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "end": 87,
          "fields": Array [
            Object {
              "end": 49,
              "fieldNumber": 4,
              "fieldOptions": Array [],
              "fieldType": Object {
                "end": 39,
                "start": 33,
                "type": "ValueType",
                "value": "string",
              },
              "name": "name",
              "start": 33,
              "type": "Field",
            },
            Object {
              "end": 83,
              "fieldNumber": 9,
              "fieldOptions": Array [],
              "fieldType": Object {
                "end": 66,
                "start": 56,
                "type": "ValueType",
                "value": "SubMessage",
              },
              "name": "sub_message",
              "start": 56,
              "type": "Field",
            },
          ],
          "oneofName": "foo",
          "options": Array [],
          "start": 15,
          "type": "OneOf",
        },
      ],
      "end": 89,
      "name": "M",
      "start": 1,
      "type": "Message",
    },
  ],
  "end": 97,
  "start": 1,
  "syntax": undefined,
  "type": "ProtoFile",
}
`;

exports[`Parser parse option 1`] = `
Object {
  "body": Array [
    Object {
      "end": 41,
      "name": Object {
        "end": 20,
        "start": 8,
        "type": "OptionName",
        "value": "java_package",
      },
      "start": 1,
      "type": "Option",
      "value": Object {
        "end": 40,
        "start": 23,
        "type": "Literal",
        "value": "com.example.foo",
      },
    },
    Object {
      "end": 67,
      "name": Object {
        "end": 61,
        "start": 49,
        "type": "OptionName",
        "value": "java_package",
      },
      "start": 42,
      "type": "Option",
      "value": Object {
        "end": 65,
        "start": 64,
        "type": "Literal",
        "value": -1,
      },
    },
    Object {
      "end": 108,
      "name": Object {
        "end": 87,
        "start": 75,
        "type": "OptionName",
        "value": "java_package",
      },
      "start": 68,
      "type": "Option",
      "value": Object {
        "end": 107,
        "start": 90,
        "type": "Literal",
        "value": "com.example.foo",
      },
    },
    Object {
      "end": 152,
      "name": Object {
        "end": 131,
        "start": 116,
        "type": "OptionName",
        "value": "(java._package)",
      },
      "start": 109,
      "type": "Option",
      "value": Object {
        "end": 151,
        "start": 134,
        "type": "Literal",
        "value": "com.example.foo",
      },
    },
    Object {
      "end": 191,
      "name": Object {
        "end": 170,
        "start": 160,
        "type": "OptionName",
        "value": "java.pp.cc",
      },
      "start": 153,
      "type": "Option",
      "value": Object {
        "end": 190,
        "start": 173,
        "type": "Literal",
        "value": "com.example.foo",
      },
    },
  ],
  "end": 192,
  "start": 1,
  "syntax": undefined,
  "type": "ProtoFile",
}
`;

exports[`Parser parse option field 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "end": 43,
          "name": Object {
            "end": 35,
            "start": 22,
            "type": "OptionName",
            "value": "(my_option).a",
          },
          "start": 15,
          "type": "Option",
          "value": Object {
            "end": 42,
            "start": 38,
            "type": "FullIdentifier",
            "value": "true",
          },
        },
      ],
      "end": 45,
      "name": "M",
      "start": 1,
      "type": "Message",
    },
  ],
  "end": 65,
  "start": 1,
  "syntax": undefined,
  "type": "ProtoFile",
}
`;

exports[`Parser parse package 1`] = `
Object {
  "body": Array [
    Object {
      "end": 11,
      "name": Object {
        "end": 10,
        "start": 9,
        "type": "FullIdentifier",
        "value": "a",
      },
      "start": 1,
      "type": "Package",
    },
    Object {
      "end": 25,
      "name": Object {
        "end": 24,
        "start": 20,
        "type": "FullIdentifier",
        "value": "a.bc",
      },
      "start": 12,
      "type": "Package",
    },
  ],
  "end": 30,
  "start": 1,
  "syntax": undefined,
  "type": "ProtoFile",
}
`;

exports[`Parser parse reserved 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "end": 26,
          "fieldName": undefined,
          "ranges": Array [
            Array [
              1,
            ],
          ],
          "start": 15,
          "type": "Reserved",
        },
        Object {
          "end": 58,
          "fieldName": undefined,
          "ranges": Array [
            Array [
              1,
              2,
            ],
            Array [
              5,
              "max",
            ],
            Array [
              6,
            ],
          ],
          "start": 29,
          "type": "Reserved",
        },
        Object {
          "end": 76,
          "fieldName": Array [
            "foo",
          ],
          "ranges": undefined,
          "start": 61,
          "type": "Reserved",
        },
        Object {
          "end": 101,
          "fieldName": Array [
            "foo",
            "bar",
          ],
          "ranges": undefined,
          "start": 79,
          "type": "Reserved",
        },
      ],
      "end": 103,
      "name": "M",
      "start": 1,
      "type": "Message",
    },
  ],
  "end": 117,
  "start": 1,
  "syntax": undefined,
  "type": "ProtoFile",
}
`;

exports[`Parser parse syntax 1`] = `
Object {
  "body": Array [],
  "end": 18,
  "start": 0,
  "syntax": Object {
    "end": 18,
    "start": 0,
    "type": "SyntaxStatement",
    "value": "proto2",
  },
  "type": "ProtoFile",
}
`;
